torch/csrc/jit/tensorexpr/cpp_codegen.cpp

```diff
@@ -1,44 +1,650 @@
+
 #include <torch/csrc/jit/tensorexpr/cpp_codegen.h>
+#include <torch/csrc/jit/tensorexpr/cpp_intrinsics.h>
+#include <torch/csrc/jit/tensorexpr/external_functions_registry.h>
+#include <torch/csrc/jit/tensorexpr/types.h>
 
 namespace torch {
 namespace jit {
 namespace tensorexpr {
 
-void CppPrinter::visit(const Allocate* alloc) {
-  constexpr size_t kAllocOnStackThresholdSize = 512;
+// Rewrites the variables' name according to valid C++ naming convention.
+// E.g. in Graph IR, variable name may contain '.', in C++, they are replaced
+// with '_'.
+class CppVarNameRewriter : public IRMutator {
+ public:
+  const Expr* mutate(const Var* v) override {
+    constexpr char kDot = '.';
+    constexpr char kUnderscore = '_';
+    if (v->name_hint().find(kDot) == std::string::npos) {
+      return v;
+    }
+    std::string name = v->name_hint();
+    std::replace(name.begin(), name.end(), kDot, kUnderscore);
+    const Var* new_v = new Var(name, v->dtype());
+    old_to_new_var_[v] = new_v;
+    return static_cast<const Expr*>(new_v);
+  }
+
+  const Expr* mutate(const Buf* v) override {
+    const Var* var = v->base_handle();
+    const Var* var_new = static_cast<const Var*>(var->accept_mutator(this));
+    if (var_new == var) {
+      return v;
+    }
+    return new Buf(var_new, v->dims(), v->dtype(), v->initializer());
+  }
+
+  // After mutation of the TE IR tree, the old Vars may still
+  // exist in CodeGen's BufferArgs, this method can be used
+  // to map the old Var to the mutated new Var.
+  const Var* getNewVar(const Var* old) const {
+    if (old_to_new_var_.find(old) != old_to_new_var_.end()) {
+      return old_to_new_var_.at(old);
+    }
+    return old;
+  }
+
+ private:
+  std::unordered_map<const Var*, const Var*> old_to_new_var_;
+};
+
+class ExprVector : public Expr {
+ public:
+  explicit ExprVector(const std::vector<const Expr*>& exprs)
+      : Expr(exprs[0]->dtype()), exprs_(exprs) {}
+
+  size_t size() const {
+    return exprs_.size();
+  }
+
+  const Expr* operator[](size_t idx) const {
+    return exprs_[idx];
+  }
+
+  void accept(IRVisitor*) const override {}
+  const Expr* accept_mutator(IRMutator*) const override {
+    return nullptr;
+  }
+
+ private:
+  std::vector<const Expr*> exprs_;
+};
+
+class StmtVector : public Stmt {
+ public:
+  explicit StmtVector(const std::vector<const Stmt*>& stmts) : stmts_(stmts) {}
+
+  size_t size() const {
+    return stmts_.size();
+  }
+
+  const Stmt* operator[](size_t idx) const {
+    return stmts_[idx];
+  }
+
+  void accept(IRVisitor*) const override {}
+  Stmt* accept_mutator(IRMutator*) override {
+    return nullptr;
+  }
+
+ private:
+  std::vector<const Stmt*> stmts_;
+};
+
+// Unrolls vector expressions into a vector of scalar expressions.
+// For example:
+// Ramp(IntImm(0), IntImm(1), IntImm(3)) is unrolled into:
+// ExprVector({IntImm(0), Add(IntImm(0), IntImm(1)), Add(IntImm(0),
+// Mul(IntImm(2), IntImm(1)))}).
+class Devectorizer : public IRMutator {
```

Comment:
Mm, I'm not sure introducing new ExprNode and Devectorizer can be justified by a single backend (c++ codegen) - generally we should avoid adding new concepts into the core that have only one use. Would it be possibleto just print statements with `Ramp` as several statements next to each other inside C++ codegen?

Reply:
I was considering a case like `Store(x, Ramp(0, 1, 3)) = Load(y, Ramp(0, 1, 3) + Ramp(0, 1, 3))`. In this case, when we visit `Store`, it's easy to expand it as 3 `Store` statements, but how can we handle the right side `Load` expression? We need a way to recursively expand the `Load` expression, and correspond the expanded expressions to the expanded `Store` statements. `Devectorizer` is the cleanest solution I came up with, do you have other ideas?

Reply:
Regarding to the new ExprNode and Devectorizer, since they are restricted to the cpp_codegen implementation, I'm less concerned about introducing these new abstractions, they can be viewed as implementation details?

Reply:
Another design is similar to eval.cpp and llvm_codegen.cpp: every time a vectorized Expr is visited, save the devectorized result into a `Value`. In this way, the logic abstracted in `Devectorizer` is spread among the `CppPrinter`'s `visit` methods. I think `Devectorizer` is a cleaner abstraction.

Reply:
Both `eval` and `llvm_codegen` natively support vectorized statements (see e.g. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/eval.h#L50-L51) - i.e. they are not deconstructing the IR into scalar version before visiting it, and I believe C++ backend should do the same.

> In this case, when we visit Store, it's easy to expand it as 3 Store statements, but how can we handle the right side Load expression?

As we're visiting a `Store` we could check whether the index is multilane and if so generate several C++ statements for it - I don't think there is a need in mutating IR for that. The codegen class might have a member indicating which lane is being generated now and visit the rhs N times (N is the number of lanes).

Reply:
Yes, the `eval` way is what I described in the above comment: https://github.com/pytorch/pytorch/pull/52980#discussion_r589181975.

Reply:
Make sense, we can DFS the IR tree for N times (N is the number of lanes). With Devectorizer, we DFS the original IR tree once and collect the devectorized results on backtracking, but we need to iterate through the N devectorized statements anyway.